util/src/main/java/com/coderanch/util/cli/InputUtility.java

```diff
@@ -0,0 +1,267 @@
+     */
+    private static final List<String> NO_SYNONYMS = List.of("n", "no");
+
+    /**
+     * The reader to read user responses with.
+     */
+    private final BufferedReader reader;
+
+    /**
+     * Constructs a new input utility that reads from {@link System#in} using the UTF-8 encoding.
+     * Data from the underlying stream will be buffered.
+     * Do not use the same stream in a different utility.
+     */
+    public InputUtility() {
+        this(System.in, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Constructor that takes an input stream to be used.
+     * Data from the underlying stream will be buffered.
+     * Do not use the same stream in a different utility.
+     *
+     * @param inputStream input stream to be used.
+     * @param charset     charset to be used.
+     */
+    public InputUtility(InputStream inputStream, Charset charset) {
+        requireThat("inputStream", inputStream, is(notNullValue()));
+        requireThat("charset", charset, is(notNullValue()));
+        reader = new BufferedReader(new InputStreamReader(inputStream, charset));
+    }
+
+    /**
+     * Displays the prompt and reads the input stream, returning a validated String.
+     *
+     * @param prompt          the prompt to display to the user.
+     * @param stringPredicate the predicate to use for validation.
+     * @return a validated string.
+     * @throws IOException if there's a problem while reading from the underlying stream.
+     */
+    public String nextString(String prompt, Predicate<? super String> stringPredicate) throws IOException {
+        System.out.println(prompt);
+        String line;
+        for (
+            line = reader.readLine();
+            !stringPredicate.test(line);
+            line = reader.readLine()
+        ) {
+            System.out.println("Invalid input.");
+        }
+
+        return line;
+    }
+
+    /**
+     * Displays the prompt and reads the input stream, returning a validated integer.
+     *
+     * @param prompt       the prompt to display to the user.
+     * @param intPredicate the predicate to use for validation.
+     * @return a validated integer.
+     * @throws IOException when there's a problem with {@link InputStream}
+     */
+    public int nextInt(String prompt, Predicate<? super Integer> intPredicate) throws IOException {
+        System.out.println(prompt);
+        String line;
+        OptionalInt num;
+        for (
+            line = reader.readLine(), num = tryIntParse(line, intPredicate);
+            num.isEmpty();
+            line = reader.readLine(), num = tryIntParse(line, intPredicate)
+        ) {
+            System.out.println("Invalid input.");
+        }
+        return num.getAsInt();
+    }
+
+    /**
+     * Tries to parse an Integer from the string and validate it.
+     *
+     * @param line         the string to parse.
+     * @param intPredicate the predicate used to validate the Integer.
```

Comment:
Write prose unless you have a good reason to refer to the Java type specifically.
```suggestion
     * @param intPredicate the predicate used to validate the integer.
```